File: app/src/main/java/com/jwu5/giphyapp/network/GiphyRequestService.java
package com.jwu5.giphyapp.network;

import com.jwu5.giphyapp.network.model.Datum;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Jiawei on 8/29/2017.
 */
public class GiphyRequestService {
    private final GiphyNetworkRequest mGiphyNetworkRequest;

    public GiphyRequestService(GiphyNetworkRequest giphyNetworkRequest) {
        mGiphyNetworkRequest = giphyNetworkRequest;
    }

    public void getTrending(final ResponseCallback callback, int limit, int offset) {
        makeNetworkCall(mGiphyNetworkRequest.getTrending(limit, offset), callback);
    }

    public void getSearchedResults(final ResponseCallback callback ,String query, int limit, int offset) {
        makeNetworkCall(mGiphyNetworkRequest.getSearchedGifs(query, limit, offset), callback);
    }

    private void makeNetworkCall(Observable<Datum> observable, final ResponseCallback callback) {
        observable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Datum>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                        callback.onSubscribe();
                    }
                    @Override
                    public void onNext(@NonNull Datum datum) {
                        callback.onSuccess(datum);
                    }
                    @Override
                    public void onError(@NonNull Throwable e) {
                        callback.onError(e);
                    }
                    @Override
                    public void onComplete() {
                        callback.onComplete();
                    }
                });
    }

    public interface ResponseCallback {
        void onSubscribe();

        void onSuccess(Datum datum);

        void onError(Throwable e);

        void onComplete();
    }

}

File: app/src/main/java/com/jwu5/giphyapp/network/GiphyRequestService.kt
package com.jwu5.giphyapp.network

import com.jwu5.giphyapp.network.model.Datum

import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by Jiawei on 8/29/2017.
 */
class GiphyRequestService(private val mGiphyNetworkRequest: GiphyNetworkRequest) {

    fun getTrending(callback: ResponseCallback, limit: Int, offset: Int) {
        makeNetworkCall(mGiphyNetworkRequest.getTrending(limit, offset), callback)
    }

    fun getSearchedResults(callback: ResponseCallback, query: String?, limit: Int, offset: Int) {
        makeNetworkCall(mGiphyNetworkRequest.getSearchedGifs(query, limit, offset), callback)
    }

    private fun makeNetworkCall(observable: Observable<Datum>, callback: ResponseCallback) {
        observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(object : Observer<Datum> {
            override fun onSubscribe(@NonNull d: Disposable) {
                callback.onSubscribe()
            }

            override fun onNext(@NonNull datum: Datum) {
                callback.onSuccess(datum)
            }

            override fun onError(@NonNull e: Throwable) {
                callback.onError(e)
            }

            override fun onComplete() {
                callback.onComplete()
            }
        })
    }

    interface ResponseCallback {
        fun onSubscribe()

        fun onSuccess(datum: Datum)

        fun onError(e: Throwable)

        fun onComplete()
    }

}
